### ISSUE TITLE
Stop a build that is known to fail (since hydra failed to build some dependencies)

### ISSUE THREAD

**Issue by tobiasBora** [NixOS/nix#7722]
**Is your feature request related to a problem? Please describe.**

I'm always frustrated when I try to upgrade my system, and realize that I spent 1 hour to build a system that was **known to fail**  to build in advance as hydra already failed to build one dependency before. This arrived to me like 3 times in a couple of days (first sagemath was failing to build on unstable, then darktable, then libreoffice…).

**Describe the solution you'd like**

**Before** starting to build anything, first check if hydra reports no failed build for the derivation that the system is going to produce. If it finds one, either fail or prints a warning (ideally with an option to toggle between both behaviors).

A separate command to check it without running a full build could also be useful (for instance to automatically find a point in unstable that can build my system correctly).

**Bonus:** It would also be great to list in this same command all the builds that will be done locally because they are not found in the binary cache or because they were cancelled on hydra… and even better to **provide an estimation of the build time** to compile it [based on the time took by hydra to compile previous versions](https://hydra.nixos.org/build/207342905) (for instance if for some reasons libreoffice's build has been cancelled on hydra, I don't want to build it locally, I would prefer to switch to another commit: seing that the estimated time to build my system is 30mn would be a strong indication the /something/ is going wrong). This could also be really useful also to find commits that don't target too much recompilation can be used to bisect a regression bug. 

**Describe alternatives you've considered**
Try to manually read the log to see some strange recompilations that should not occur… unfortunately this is mostly based on experience (is it normal if wpa_supplicant is recompiled?), and sometimes the system may spend like 30mn to build stuff that I must anyway build (e.g. local software derivations, proprietary virtualbox extensions…), and only fail later on a really buggy recompilation.

**Priorities**

Add :+1: to [issues you find important](https://github.com/NixOS/nix/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).

**Comment by NickCao** [comment]
Sounds like a *negative* binary cache can fulfill most of you needs.
And there might had been such a thing:
https://github.com/NixOS/nix/blob/0079d2943702a7a7fbdd88c0f9a5ad677c334aa8/src/libstore/worker-protocol.hh#L38

**Comment by tobiasBora** [comment]
Exactly, a negative binary cache seems to be all we need to provide the necessary information (of course, nix needs to be upgraded to use this information). It's nice that this might have been created before, I don't know enough of hydra to know if it uses it or not ^^'

**Comment by NickCao** [comment]
Hydra does cache failed builds itself, but is not exposing these information via binary caches. Nix used to remember failed builds, but this functionality was later removed for reasons I do not know.

**Comment by NickCao** [comment]
Found https://github.com/NixOS/nix/commit/8cffec848
